design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,352 @@
+occurrence of the key column in the projection as naming the key column in the result, and any
+subsequent copies to be value columns, (see example below). However, this reliance on the _order_
+of columns is less than ideal. Anyone got any better ideas for a short term fix for this edge case?
+
+```sql
+CREATE TABLE OUTPUT AS SELECT ID, V0, V1, ID AS V2 FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT
+
+
+CREATE TABLE OUTPUT AS SELECT ID AS NEW_KEY, V0, V1, ID FROM INPUT;
+-- resulting schema: NEW_KEY INT KEY, V0 INT, V1 INT, ID INT
+```
+
+## What is in scope
+
+- Removal of implicit copying of key column,
+  in favour of requiring key column in projection of persistent queries.
+- Removal of non-standard GROUP BY, PARTITION BY and JOIN aliasing syntax,
+  in favour of standard aliasing of the key column in the projection.
+- removal of duplicate left join column on 'select *' joins.
+- TBD: syntax for allowing user to add key column to value schema.
+
+## What is not in scope
+
+- Changes in syntax for changing the key column, e.g. allowing the projection to change the key
+  column.  This out of scope and is only potential future work. It should be discussion separately.
+- Replacing the use of the `KEY` keyword in CSAS statements with other syntax:
+  this is mostly orthogonal to this change.
+- everything else.
+
+## Value/Return
+
+Standardizing the key semantics in queries will lower the barrier for entry for users and engineers
+alike, and reduce the support burden of explaining the subtleties. It will also simplify the code,
+which should result in less bugs.
+
+Removing of the non-standard `GROUP BY`, `PARTITION BY` and `JON ON` aliasing in favour of aliasing
+in the projection will improve out standards compliance.
+
+## Public APIS
+
+1. Persistent queries, i.e. those used in `CREATE TABLE AS`, `CREATE STREAM AS` and `INSERT INTO`
+statements, will be required to _always_ include their key columns in their projection. An error
+will be generated should the projection of a persistent query not include its key column. For
+example:
+
+```sql
+-- old syntax that worked:
+CREATE TABLE OUTPUT AS SELECT COUNT() AS COUNT FROM INPUT GROUP BY V0;
+-- will now fail with an error explaining the projection must include the key column `V0`.sql
+
+-- corrected query:
+CREATE TABLE OUTPUT AS SELECT V0, COUNT() AS COUNT FROM INPUT GROUP BY V0;
+-- resulting schema: V0 INT KEY, COUNT BIGINT
+```
+
+2. The, as yet unreleased, non-standard `GROUP BY`, `PARTITION BY` and `JOIN ON` alias syntax will
+be removed in favour of using the existing standard-compliant aliasing in the projection. For
+example:
+
+```sql
+-- 'any key' aliasing syntax that will be dropped:
+CREATE TABLE OUTPUT AS SELECT COUNT() AS COUNT FROM INPUT GROUP BY (V0 + V1) AS K;
+-- resulting schema: K INT KEY, COUNT BIGINT
+
+-- proposed key column aliasing in projection:
+CREATE TABLE OUTPUT AS SELECT (V0 + v1) AS K, COUNT() AS COUNT FROM INPUT GROUP BY V0 + V1;
```

Comment:
This is actually a mistake in the KLIP - aliasing for multiple join columns will not be supported.

As to you're question: no it won't as the `GROUP BY` is applied to `INPUT` _before_ the projection, hence `K` does not yet exist.

Reply:
@big-andy-coates can we just swap in the expression associated with `K` before running the query? This will work in most SQL systems.

Reply:
What if `INPUT` has a column named `K`?  It would then be ambiguous as to which column the user was referring to in the grouping clause. 

So... no, that doesn't work.

Reply:
We could just throw an ambiguous column reference error in that particular case. Here's an example of how this works in Postgres:

```
postgres=# \d test
                Table "public.test"
 Column |  Type   | Collation | Nullable | Default 
--------+---------+-----------+----------+---------
 x      | integer |           |          | 

postgres=# select x+x as g, count(*) from test group by g;
 g  | count 
----+-------
  4 |     2
 10 |     2
  6 |     2
 14 |     2
  2 |     2
 16 |     2
 12 |     2
 20 |     2
 18 |     2
  8 |     2
(10 rows)
```
Now add a column `g`:
```
postgres=# alter table test add column g integer;
ALTER TABLE
postgres=# \d test
                Table "public.test"
 Column |  Type   | Collation | Nullable | Default 
--------+---------+-----------+----------+---------
 x      | integer |           |          | 
 g      | integer |           |          | 

postgres=# select x+x as g, count(*) from test group by g;
ERROR:  column "test.x" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: select x+x as g, count(*) from test group by g;
```

Reply:
Good to know - thanks.